src/query/list_file.c

```diff
@@ -363,20 +366,31 @@ qfile_list_cache_cleanup (THREAD_ENTRY * thread_p)
 		      // exclude in-transaction
 		      continue;
 		    }
-		  page_ref = candidate.qcache->list_id.page_cnt / (candidate.qcache->ref_count + 1);
-		  lru_sec = current_time.tv_sec - candidate.qcache->time_last_used.tv_sec;
-		  clr_cnt = candidate.qcache->xcache_entry->clr_count + 1;
-		  candidate.weight = page_ref * lru_sec * clr_cnt;
+		  page_ref = candidate.qcache->list_id.page_cnt + 1;
+		  lru_sec = current_time.tv_sec - candidate.qcache->time_last_used.tv_sec + 1;
+		  candidate.weight = (double) (candidate.qcache->ref_count + 1) / (double) (page_ref * lru_sec);
 		  (void) bh_try_insert (bh, &candidate, NULL);
 		}
 	    }
 	}
     }
 
-  for (candidate_index = 0; candidate_index < bh->element_count; candidate_index++)
+  /* traverse in reverse for weight ordering, from light weight to heavy weight */
+  for (candidate_index = bh->element_count - 1; candidate_index >= 0; candidate_index--)
     {
       bh_element_at (bh, candidate_index, &candidate);
+
+      invalidate = candidate.qcache->invalidate;
+      ht_no = candidate.qcache->list_ht_no;
       qfile_delete_list_cache_entry (thread_p, candidate.qcache);
+      if (invalidate && qfile_get_list_cache_number_of_entries (ht_no) == 0)
```

Comment:
Deletion of qcache entries proceeds without checking last_ta_idx, deletion_marker, invalidate, CSECT_QPROC_LIST_CACHE.
qfile_end_use_of_list_cache_entry() checks the following.
```
  if (lent->last_ta_idx == 0)
    {
      /* to check if it's the last transaction using the lent */
      if (marker || lent->deletion_marker)
	{
	  invalidate = lent->invalidate;
	  qfile_delete_list_cache_entry (thread_p, lent);
	}
    }
```

Reply:
lent->last_ta_idx is already checked at victim selection.